### PR TITLE
Add has / !has operators

### PR DESCRIPTION
**Pull request by lucaswoj**
ref https://github.com/mapbox/mapbox-gl-style-spec/issues/364

This will require a minor version bump. 

cc @mourner @jfirebaugh @ansis 

# Todo

 - [ ] merge this PR
 - [ ] release new `feature-filter` version
 - [ ] add support to `mapbox-gl-style-spec`
 - [ ] update `feature-filter` and `mapbox-gl-style-spec` in GL JS
 - [ ] port changes to GL Native

### CODE REVIEW

**Comment by mourner** [comment]
:+1:

File: test.js [hunk 1]
```diff
@@ -345,3 +391,29 @@ test('none', function(t) {
...
+
+test('has', function(t) {
+    var f = filter(['has', 'foo']);
+    t.equal(f({properties: {foo: 0}}), true);
+    t.equal(f({properties: {foo: 1}}), true);
+    t.equal(f({properties: {foo: '0'}}), true);
+    t.equal(f({properties: {foo: true}}), true);
+    t.equal(f({properties: {foo: false}}), true);
+    t.equal(f({properties: {foo: null}}), true);
+    t.equal(f({properties: {foo: undefined}}), false);
```
Comment (jfirebaugh) [inline test.js]: I'd expect this to be `true`.

File: test.js [hunk 2]
```diff
@@ -345,3 +391,29 @@ test('none', function(t) {
...
+
+test('!has', function(t) {
+    var f = filter(['!has', 'foo']);
+    t.equal(f({properties: {foo: 0}}), false);
+    t.equal(f({properties: {foo: 1}}), false);
+    t.equal(f({properties: {foo: '0'}}), false);
+    t.equal(f({properties: {foo: false}}), false);
+    t.equal(f({properties: {foo: false}}), false);
+    t.equal(f({properties: {foo: null}}), false);
+    t.equal(f({properties: {foo: undefined}}), true);
```
Comment (jfirebaugh) [inline test.js]: And this to be `false`.

**Comment by lucaswoj** [comment]
This PR is blocked on a decision about how the `has` operator should handle "literally undefined", `undefined`, and `null`. In vector tiles, the three are one-in-the-same. In GeoJSON, the three could be differentiated.

Open questions:

 - Should `has` evaluate to true or false for `null`?
 - Should `has` evaluate to true or false for `undefined`?
 - Should `has` evaluate to true or false for "literally undefined"?
 - Will these decisions be consistent with GL Native? 
 - Will these behaviours be predictable for a GeoJSON file converted to vector tiles serverside? 
 - Will these decisions be consistent with future implementations of geojson-vt?

**Comment by jfirebaugh** [comment]
Let's go with the definitions you've chosen:

* `has` is false for `undefined`
* `has` is false for "literally undefined"
* `has` is true for `null`

**Comment by arunasank** [comment]
:wave: @lucaswoj - this is going to be super useful to find stats with https://github.com/mapbox/osm-tag-stats (https://github.com/mapbox/osm-tag-stats/issues/2) . Will this get merged soon?

**Comment by lucaswoj** [comment]
@arunasank I'll get this into the next release!

**Comment by lucaswoj** [comment]
With humility and hindsight, I think @jfirebaugh's proposed definitions for `undefined` are best. This is ready for 👀 . I also added a commit with some minor feature filter cleanup.

File: index.js [hunk 3]
```diff
@@ -58,6 +66,14 @@ function compileIn(key, values) {
...
+function compileHasOp(property) {
+    return JSON.stringify(property) + ' in p';
+}
+
+function compileNegation(expression) {
+    return '!(' + expression + ')';
+}
+
+function compareDescending(a, b) {
+    return a - b;
```
Comment (jfirebaugh) [inline index.js]: This doesn't work for strings.

**Comment by jfirebaugh** [comment]
> With humility and hindsight, I think @jfirebaugh's proposed definitions for `undefined` are best.

Ah, ok, I'm fine with that too. 😄  For those following along, the behavior in question is "should `has` evaluate to true or false for a property that exists and whose value is `undefined`?" The behavior we're settling on is that `has` will have the same behavior as the Javascript `in` operator, evaluating to true in that case.

File: index.js (hunk 3 above)
Comment (lucaswoj) [inline index.js]: Fixed in latest commit

**Comment by lucaswoj** [comment]
Cleaned up the commit history and addressed https://github.com/mapbox/feature-filter/pull/15#discussion_r61961534. Ready to 🚢?

File: index.js (hunk 3 above)
Comment (jfirebaugh) [inline index.js]: How so? I still see `a - b` which returns `NaN` for strings.

Comment (lucaswoj) [inline index.js]: 😞 It got lost in the rebase shuffle. 😄 Fixed now!

**Comment by jfirebaugh** [comment]
🚢

**Pull request merged**